### DISCUSSION TITLE
Maximum lambda size across pages/apis

### DISCUSSION THREAD

**Discussion by sambecker** [vercel/vercel#4013] [General] [answered]
I've run into an issue where I can either deploy a page (typical size `10mb`) or an API endpoint from the `pages/api` directory (typical size `40mb`) but not both. If I delete the page, the api will deploy. If I delete the api, the page will deploy.

```
Error: The Serverless Function "stories/[storyId]" is 51.94mb which exceeds the maximum size limit of 50mb. . Learn more: https://zeit.co/docs/v2/platform/limits#serverless-function-size
```

It seems as if these two lambdas contribute toward the same file size quota.

Can anyone confirm if that's true? I had thought that dependencies/static assets were split across lambdas and that they all had some theoretical maximum footprint.

If there is a collective limit, is that limit the same across all Zeit pricing tiers?

_It's worth noting here that my API contains `chrome/puppeteer` which clearly adds some heft. However, it deploys fine if I don't deploy a `page` alongside it._

FWIW these are my deploy logs:

```
23:43:36.773  Page                                                           Size     First Load JS
23:43:36.773  ┌ ○ /                                                          6.92 kB        64.2 kB
23:43:36.773  ├   /_app                                                      307 B          57.3 kB
23:43:36.773  ├ ○ /[userId]                                                  270 B          57.6 kB
23:43:36.773  ├ ○ /404                                                       3.15 kB        60.5 kB
23:43:36.773  ├ λ /api/stories/[message]
23:43:36.774  └ λ /stories/[storyId]                                         235 kB          293 kB
23:43:36.774  + First Load JS shared by all                                  57.3 kB
23:43:36.774    ├ static/pages/_app.js                                       307 B
23:43:36.774    ├ chunks/db6f4259d89e74190e55ae87b1fd2e84d5748550.627b6b.js  10.1 kB
23:43:36.774    ├ chunks/framework.0f140d.js                                 40 kB
23:43:36.774    ├ runtime/main.3fc2ec.js                                     6.2 kB
23:43:36.774    ├ runtime/webpack.b65cab.js                                  746 B
23:43:36.774    └ css/fd8a2041f4f19a749b86.css                               22.9 kB
23:43:36.774  λ  (Lambda)  server-side renders at runtime (uses getInitialProps or getServerSideProps)
23:43:36.774  ○  (Static)  automatically rendered as static HTML (uses no initial props)
23:43:36.774  ●  (SSG)     automatically generated as static HTML + JSON (uses getStaticProps)
23:43:36.835  Done in 16.68s.
23:43:41.484  Traced Next.js serverless functions for external files in: 4632.590ms
23:43:51.306  Compressed shared serverless function files: 9820.605ms
23:43:51.439  All serverless functions created in: 131.296ms
23:43:51.560  Error: The Serverless Function "stories/[storyId]" is 51.94mb which exceeds the maximum size limit of 50mb. . Learn more: https://zeit.co/docs/v2/platform/limits#serverless-function-size
23:43:53.123  Done with "package.json"
```

**Comment by sambecker** [comment]
Does anyone have any idea how to manually shrink a lambda? I've tried restricting dependencies on a function-by-function basis but it doesn't seem to have an effect.

Any idea what I'm going wrong?

from `now.json`:

```
{
  "functions": {
    "pages/api/stories/image/[storyId].ts": {
      "excludeFiles": "node_modules/firebase"
    },
    "pages/api/stories/[storyId].ts": {
      "excludeFiles": "node_modules/chrome-aws-lambda/**"
    },
    "pages/stories/[storyId].ts": {
      "excludeFiles": "node_modules/chrome-aws-lambda/**"
    }
  }
}
```

**Comment by paulogdm** [comment]
Are you able to create a minimal reproducible example?

**Comment by sambecker** [comment]
@paulogdm just created a repo!

https://github.com/sambecker/now-function-size

**Comment by styfle** [comment, marked as answer]
Thanks for the [repo](https://github.com/sambecker/now-function-size)!

This will be fixed in PR #4018.

Also, please update to the latest Next.js version to ensure the latest bug fixes.

**Reply by sambecker** [reply]
Thank you @styfle + @paulogdm!

**Reply by styfle** [reply]
This has been fixed. I confirmed that your `now-function-size` repo above is no longer failing to build.

**Reply by sambecker** [reply]
Thank you so much @styfle + @paulogdm!

**Comment by sambecker** [comment]
Hi @styfle + @paulogdm just ran into a similar issue!

I created a new API endpoint that used the same dependency as my original page (in this case `firebase`) and ran into the same size constraint error. It seems like all API dependencies are added together, the same way that page dependencies were previously?

Is that intended behavior? Does each API become its own discrete lambda (or do all pages get pooled together for a `50mb` limit and all APIs get pooled together for an independent `50mb` limit)?

I've updated my example repo to demonstrate,

https://github.com/sambecker/now-function-size

**Comment by sambecker** [comment]
To be clear, the issue I'm having now: is in order for my app to deploy I need to delete either:

- `test-api.ts` or
- `test-api-alt.ts`

Either of those will deploy if the other has been removed.

As with the issue I was having before, it seems like all dependencies for all API endpoints are being bundled together versus being intelligently split up per lambda.

Any help would be greatly appreciated!

**Comment by sambecker** [comment]
Any thoughts on what might be causing this @paulogdm? Should I re-post this as a new thread?

**Reply by ChrisLusted** [reply]
@sambecker did you get anywhere with this? 

I'm having a similar issue on my project and also with a separate test deploy using your repo

**Reply by sambecker** [reply]
I'm not sure @ChrisLusted eventually my project deployed but it may have been because Vercel increased the size of permitted lambdas? I believe the problem of API dependencies polluting page dependencies went away. I'm not sure if one api lambdas size constraint is impacted by others.

TDLR; my project eventually became deployable but I'm not exactly sure why. Frankly, I can't even find the size of my lambdas reported on Vercel's dashboard!

**Reply by gregavola** [reply]
Having the same issue here, my functions are 50MB a piece, with no content in them, just an empty file. Are these shares some how?

**Reply by flogst** [reply]
I'm having the exact same issue as well. @gregavola @ChrisLusted did you find any solution in the end?

**Reply by sambecker** [reply]
And mine no longer deploy on `10.0.8` which might be due to it being a heavier version of `Next`? Anyway, I'm sticking with `10.0.7` until something changes. Wish lambda sizes were easier to debug!

**Reply by DrunkOldDog** [reply]
Same problem here!

**Comment by styfle** [comment]
@sambecker I created a PR to your repo here: https://github.com/sambecker/now-function-size/pull/11

cc @marklawlor who had a similar problem
